iot/api-client/mqtt_example/cloudiot_mqtt_example.py

```diff
@@ -169,37 +169,200 @@ def get_client(
+    time.sleep(5)
+
+    # The topic devices receive configuration updates on.
+    device_config_topic = '/devices/{}/config'.format(device_id)
+    client.subscribe(device_config_topic, qos=1)
+
+    # The topic gateways receive configuration updates on.
+    gateway_config_topic = '/devices/{}/config'.format(gateway_id)
+    client.subscribe(gateway_config_topic, qos=1)
+
+    # The topic gateways receive error updates on. QoS must be 0.
+    error_topic = '/devices/{}/errors'.format(gateway_id)
+    client.subscribe(error_topic, qos=0)
+
+    # Wait for about a minute for config messages.
+    for i in range(1, duration):
+        client.loop()
+        if cb is not None:
+            cb(client)
+
+        if should_backoff:
+            # If backoff time is too large, give up.
+            if minimum_backoff_time > MAXIMUM_BACKOFF_TIME:
+                print('Exceeded maximum backoff time. Giving up.')
+                break
+
+            delay = minimum_backoff_time + random.randint(0, 1000) / 1000.0
+            time.sleep(delay)
+            minimum_backoff_time *= 2
+            client.connect(mqtt_bridge_hostname, mqtt_bridge_port)
+
+        seconds_since_issue = (datetime.datetime.utcnow() - jwt_iat).seconds
+        if seconds_since_issue > 60 * jwt_exp_mins:
+            print('Refreshing token after {}s').format(seconds_since_issue)
+            jwt_iat = datetime.datetime.utcnow()
+            client = get_client(
+                project_id, cloud_region, registry_id, gateway_id,
+                private_key_file, algorithm, ca_certs, mqtt_bridge_hostname,
+                mqtt_bridge_port)
+
+        time.sleep(1)
+
+    detach_device(client, device_id, mqtt_bridge_hostname, mqtt_bridge_port)
+
+    print('Finished.')
+    # [END listen_for_config_messages]
+
+
+def send_data_from_bound_device(
+        service_account_json, project_id, cloud_region, registry_id, device_id,
+        gateway_id, num_messages, private_key_file, algorithm, ca_certs,
+        mqtt_bridge_hostname, mqtt_bridge_port, jwt_expires_minutes, payload):
+    """Sends data from a gateway on behalf of a device that is bound to it."""
+    # [START send_data_from_bound_device]
+    global minimum_backoff_time
+
+    # Publish device events and gateway state.
+    device_topic = '/devices/{}/{}'.format(device_id, 'state')
+    gateway_topic = '/devices/{}/{}'.format(gateway_id, 'state')
+
+    jwt_iat = datetime.datetime.utcnow()
+    jwt_exp_mins = jwt_expires_minutes
+    # Use gateway to connect to server
+    client = get_client(
+        project_id, cloud_region, registry_id, gateway_id,
+        private_key_file, algorithm, ca_certs, mqtt_bridge_hostname,
+        mqtt_bridge_port)
+
+    attach_device(client, device_id, mqtt_bridge_hostname, mqtt_bridge_port)
+    print('Waiting for device to attach.')
+    time.sleep(5)
+
+    # Publish state to gateway topic
+    gateway_state = 'Starting HUB at: {}'.format(time.time())
```

Comment:
might be confusing if we use both Hub and Gateway. Might want to just change this to 'Starting gateway at...'

Reply:
done